File: Conv.c
#include <stdio.h>
#include <math.h>

// WGS84定数
#define WGS84_A 6378137.0
#define WGS84_F (1.0/298.257223563)
#define M_PI 3.14159265358979323846

static double deg2rad(double deg) {
    return deg * M_PI / 180.0;
}

// 緯度経度からECEF(X, Y, Z)変換
// lat, lon: degree, h: height [m]
static void latlon_to_ecef(double lat_deg, double lon_deg, double h,
                           double *X, double *Y, double *Z) {
    double e2 = 2.0 * WGS84_F - WGS84_F * WGS84_F;
    double lat = deg2rad(lat_deg);
    double lon = deg2rad(lon_deg);

    double sin_lat = sin(lat);
    double cos_lat = cos(lat);
    double sin_lon = sin(lon);
    double cos_lon = cos(lon);

    double N = WGS84_A / sqrt(1.0 - e2 * sin_lat * sin_lat);

    *X = (N + h) * cos_lat * cos_lon;
    *Y = (N + h) * cos_lat * sin_lon;
    *Z = (N * (1.0 - e2) + h) * sin_lat;
}

// ECEF -> ENU変換
// (X, Y, Z): ECEF座標
// (lat0_deg, lon0_deg, h0): 原点の緯度経度・高さ
// 戻り値: E, N, U (ENU座標)
static void ecef_to_enu(double X, double Y, double Z,
                        double lat0_deg, double lon0_deg, double h0,
                        double *E, double *N, double *U) {
    double X0, Y0, Z0;
    latlon_to_ecef(lat0_deg, lon0_deg, h0, &X0, &Y0, &Z0);

    double dX = X - X0;
    double dY = Y - Y0;
    double dZ = Z - Z0;

    double lat0 = deg2rad(lat0_deg);
    double lon0 = deg2rad(lon0_deg);

    double sin_lat0 = sin(lat0);
    double cos_lat0 = cos(lat0);
    double sin_lon0 = sin(lon0);
    double cos_lon0 = cos(lon0);

    // East軸単位ベクトル
    double e_x = -sin_lon0;
    double e_y =  cos_lon0;
    double e_z = 0.0;

    // North軸単位ベクトル
    double n_x = -sin_lat0 * cos_lon0;
    double n_y = -sin_lat0 * sin_lon0;
    double n_z =  cos_lat0;

    // Up軸単位ベクトル
    double u_x = cos_lat0 * cos_lon0;
    double u_y = cos_lat0 * sin_lon0;
    double u_z = sin_lat0;

    *E = e_x * dX + e_y * dY + e_z * dZ;
    *N = n_x * dX + n_y * dY + n_z * dZ;
    *U = u_x * dX + u_y * dY + u_z * dZ;
}

int main(void) {
    // 原点
    double lat0_deg = 35.0;
    double lon0_deg = 135.0;
    double h0 = 0.0;

    // 対象点(例): 北方向へ約1kmほど離れた地点
    double lat_deg = 35.0001;
    double lon_deg = 135.0;
    double h = 0.0;

    // 対象点をECEFへ
    double X, Y, Z;
    latlon_to_ecef(lat_deg, lon_deg, h, &X, &Y, &Z);

    // ECEF -> ENU
    double E, N, U;
    ecef_to_enu(X, Y, Z, lat0_deg, lon0_deg, h0, &E, &N, &U);

    printf("E = %f [m]\nN = %f [m]\nU = %f [m]\n", E, N, U);

    // 原点からの水平距離
    double distance = sqrt(E*E + N*N);
    printf("水平距離 = %f [m]\n", distance);

    return 0;
}

File: conv.py
import math
import tkinter as tk
from tkinter import messagebox
import matplotlib.pyplot as plt

# --------------------------------------------
# 定数および計算用関数
# --------------------------------------------
# WGS84の地球楕円体定数
a = 6378137.0            # 地球長半径 [m]
f = 1.0/298.257223563    # 扁平率
e2 = 2*f - f*f           # 第一離心率^2

def latlon_to_ecef(lat_deg, lon_deg, h=0.0):
    """
    緯度経度(度)と高さh(m)からECEF座標(X,Y,Z)を求める関数。
    この関数により地理座標系から地球中心固定座標系へ変換し、直交座標上でベクトル演算が可能になる。
    """
    lat = math.radians(lat_deg)
    lon = math.radians(lon_deg)
    sin_lat = math.sin(lat)
    cos_lat = math.cos(lat)
    sin_lon = math.sin(lon)
    cos_lon = math.cos(lon)
    
    N = a / math.sqrt(1 - e2*(sin_lat**2))
    X = (N + h) * cos_lat * cos_lon
    Y = (N + h) * cos_lat * sin_lon
    Z = (N*(1 - e2) + h)*sin_lat
    return X, Y, Z

def ecef_to_latlon(X, Y, Z):
    """
    ECEF(X,Y,Z)から緯度経度高さを求める関数。
    Bowring法などを用いてWGS84楕円体上での逆変換を行う。
    """
    # 経度lon
    lon = math.atan2(Y, X)
    
    p = math.sqrt(X*X + Y*Y)
    # 短半径b
    b = a * math.sqrt(1 - e2)
    # 第2離心率^2
    e2p = (a*a - b*b)/(b*b)
    
    # Bowringの方法に基づく計算
    theta = math.atan2(Z * a, p * b)
    sin_theta = math.sin(theta)
    cos_theta = math.cos(theta)
    
    phi = math.atan2(Z + e2p * b * (sin_theta**3),
                     p - e2 * a * (cos_theta**3))
    
    sin_phi = math.sin(phi)
    N = a / math.sqrt(1 - e2 * sin_phi*sin_phi)
    h = p / math.cos(phi) - N
    
    lat_deg = math.degrees(phi)
    lon_deg = math.degrees(lon)
    return lat_deg, lon_deg, h

def compute_projection(A, B, P):
    """
    A,B,P(緯度経度)から、PをA-B直線上に投影した点P'を求める関数。
    手順：
    1. A,B,PをECEF座標へ変換
    2. ベクトルAB,APを算出し、内積計算からパラメータtを求める
    3. P' = A + t*AB をECEFで求める
    4. ECEF→緯度経度逆変換でP'を求める
    """
    X_A, Y_A, Z_A = latlon_to_ecef(A[0], A[1])
    X_B, Y_B, Z_B = latlon_to_ecef(B[0], B[1])
    X_P, Y_P, Z_P = latlon_to_ecef(P[0], P[1])
    
    AB = (X_B - X_A, Y_B - Y_A, Z_B - Z_A)
    AP = (X_P - X_A, Y_P - Y_A, Z_P - Z_A)
    
    APdotAB = AB[0]*AP[0] + AB[1]*AP[1] + AB[2]*AP[2]
    ABdotAB = AB[0]*AB[0] + AB[1]*AB[1] + AB[2]*AB[2]
    
    t = APdotAB / ABdotAB
    
    X_Pp = X_A + t*AB[0]
    Y_Pp = Y_A + t*AB[1]
    Z_Pp = Z_A + t*AB[2]
    
    lat_Pp, lon_Pp, _ = ecef_to_latlon(X_Pp, Y_Pp, Z_Pp)
    return (lat_Pp, lon_Pp)

# --------------------------------------------
# GUI クラス定義
# --------------------------------------------
class App:
    def __init__(self, master):
        master.title("A-Bライン上へのP投影ツール")
        
        # A点入力
        tk.Label(master, text="A 緯度:").grid(row=0, column=0, sticky="e")
        tk.Label(master, text="A 経度:").grid(row=0, column=2, sticky="e")
        self.A_lat = tk.Entry(master)
        self.A_lon = tk.Entry(master)
        self.A_lat.grid(row=0, column=1)
        self.A_lon.grid(row=0, column=3)
        
        # B点入力
        tk.Label(master, text="B 緯度:").grid(row=1, column=0, sticky="e")
        tk.Label(master, text="B 経度:").grid(row=1, column=2, sticky="e")
        self.B_lat = tk.Entry(master)
        self.B_lon = tk.Entry(master)
        self.B_lat.grid(row=1, column=1)
        self.B_lon.grid(row=1, column=3)
        
        # P点入力
        tk.Label(master, text="P 緯度:").grid(row=2, column=0, sticky="e")
        tk.Label(master, text="P 経度:").grid(row=2, column=2, sticky="e")
        self.P_lat = tk.Entry(master)
        self.P_lon = tk.Entry(master)
        self.P_lat.grid(row=2, column=1)
        self.P_lon.grid(row=2, column=3)
        
        # デフォルト値（例）
        self.A_lat.insert(0, "35.394927")
        self.A_lon.insert(0, "136.850982")
        self.B_lat.insert(0, "35.393738")
        self.B_lon.insert(0, "136.888318")
        self.P_lat.insert(0, "35.44375")
        self.P_lon.insert(0, "136.865959")
        
        # 計算＆プロットボタン
        self.compute_button = tk.Button(master, text="計算＆プロット", command=self.compute_and_plot)
        self.compute_button.grid(row=3, column=0, columnspan=4, pady=10)
        
    def compute_and_plot(self):
        # 入力値取得
        try:
            A_lat = float(self.A_lat.get())
            A_lon = float(self.A_lon.get())
            B_lat = float(self.B_lat.get())
            B_lon = float(self.B_lon.get())
            P_lat = float(self.P_lat.get())
            P_lon = float(self.P_lon.get())
        except ValueError:
            messagebox.showerror("入力エラー", "有効な数値を入力してください。")
            return
        
        A = (A_lat, A_lon)
        B = (B_lat, B_lon)
        P = (P_lat, P_lon)
        
        # P'計算
        P_prime = compute_projection(A, B, P)
        
        # 結果表示
        msg = f"P'の計算結果:\n緯度: {P_prime[0]:.6f}\n経度: {P_prime[1]:.6f}"
        messagebox.showinfo("結果", msg)
        
        # プロット: latを縦軸、lonを横軸として単純表示
        plt.figure(figsize=(6,4))
        
        # A,B,P,P'をプロット
        lats = [A[0], B[0], P[0], P_prime[0]]
        lons = [A[1], B[1], P[1], P_prime[1]]
        colors = ["red", "blue", "green", "purple"]
        labels = ["A","B","P","P'"]
        
        for (la, lo, c, lbl) in zip(lats, lons, colors, labels):
            plt.scatter(lo, la, color=c)
            plt.text(lo, la, f" {lbl}", fontsize=9, color=c)
        
        plt.xlabel("経度(度)")
        plt.ylabel("緯度(度)")
        plt.title("A, B, P および P' の位置")
        plt.grid(True)
        plt.show()

# --------------------------------------------
# メインループ開始
# --------------------------------------------
if __name__ == "__main__":
    root = tk.Tk()
    app = App(root)
    root.mainloop()